tests/api2/pool_dataset_encryption.py

```diff
@@ -0,0 +1,870 @@
+    }
+    results = POST('/pool/dataset/', payload)
+    assert results.status_code == 422, results.text
+    assert 'Must be disabled when encryption is enabled' in results.text, results.text
+
+
+def test_12_try_to_create_an_encrypted_dataset_with_passphrase_and_generate_key():
+    payload = {
+        'name': dataset,
+        'encryption_options': {
+            'generate_key': True,
+            'passphrase': 'my_passphrase',
+        },
+        'encryption': True,
+        'inherit_encryption': False
+    }
+    results = POST('/pool/dataset/', payload)
+    assert results.status_code == 422, results.text
+    assert 'Must be disabled when dataset is to be encrypted with passphrase' in results.text, results.text
+
+
+def test_13_create_an_encrypted_root_with_generate_key():
+    payload = {
+        'name': dataset,
+        'encryption_options': {
+            'generate_key': True,
+        },
+        'encryption': True,
+        'inherit_encryption': False
+    }
+    results = POST('/pool/dataset/', payload)
+    assert results.status_code == 200, results.text
+
+
+@skip_for_ha
+def test_14_delete_generate_key_encrypted_root():
+    results = DELETE(f'/pool/dataset/id/{dataset_url}/')
+    assert results.status_code == 200, results.text
+
+
+@skip_for_ha
+def test_15_create_an_encrypted_root_with_generate_key():
+    payload = {
+        'name': dataset,
+        'encryption_options': {
+            'key': dataset_token_hex,
+        },
+        'encryption': True,
+        'inherit_encryption': False
+    }
+    results = POST('/pool/dataset/', payload)
+    assert results.status_code == 200, results.text
+
+
+@skip_for_ha
+def test_16_make_sure_we_are_not_able_to_lock_key_encrypted_dataset():
+    payload = {
+        'id': dataset,
+        'lock_options': {
+            'force_umount': True
+        }
+    }
+    results = POST('/pool/dataset/lock', payload)
+    assert results.status_code == 422, results.text
+    assert 'Only datasets which are encrypted with passphrase can be locked' in results.text, results.text
+
+
+@skip_for_ha
+def test_17_change_a_key_encrypted_dataset_to_passphrase():
+    payload = {
+        'id': dataset,
+        'change_key_options': {
+            'passphrase': 'my_passphrase'
+        }
+    }
+    results = POST('/pool/dataset/change_key/', payload)
+    assert results.status_code == 200, results.text
+    job_id = results.json()
+    job_status = wait_on_job(job_id, 120)
+    assert job_status['state'] == 'SUCCESS', str(job_status['results'])
+
+
+@skip_for_ha
+def test_18_lock_passphrase_encrypted_datasets_and_ensure_they_get_locked():
+    payload = {
+        'id': dataset,
+        'lock_options': {
+            'force_umount': True
+        }
+    }
+    results = POST('/pool/dataset/lock', payload)
+    assert results.status_code == 200, results.text
+
+
+@skip_for_ha
+def test_19_verify_passphrase_encrypted_root_is_locked():
+    payload = {
+        'id': dataset
+    }
```

Comment:
Let's please pass the `passphrase` for the dataset as well and then ensure that `unlock_successful` attribute says `True` which means that if we unlock the dataset in question, it will indeed get unlocked. It might be nice to test with a wrong passphrase as well to ensure the data reported by middleware is accurate.

Reply:
> Let's please pass the `passphrase` for the dataset as well and then ensure that `unlock_successful` attribute says `True` which means that if we unlock the dataset in question, it will indeed get unlocked.

I am not sure what you mean there is a test already for that.